### ISSUE TITLE
Error with empty summarize() with data.table

### ISSUE THREAD

**Issue by krlmlr** [hadley/dplyr#1073]
```r
library(dplyr)
```

Works with `data.frame`:

```r
iris %>% group_by(Species) %>% summarize %>% nrow
```

```
## [1] 3
```

Doesn't works with `data.table`:

```r
iris %>% tbl_dt %>% group_by(Species) %>% summarize %>% nrow
```

```
## Error in jsub[[ii]]: attempt to select less than one element
```

```r
iris %>% tbl_dt %>% group_by(Species) %>% summarize_ %>% nrow
```

```
## Error in jsub[[ii]]: attempt to select less than one element
```

Works with `data.table` if `n()` is called explicitly:

```r
iris %>% tbl_dt %>% group_by(Species) %>% summarize(n()) %>% nrow
```

```
## [1] 3
```

**Comment by romainfrancois** [comment]
So internally it's like we were calling this `data.table` code: 

```r
> iris_dt <- as.data.table(iris)
> iris_dt[ , list(), by = "Species" ]
Error in jsub[[ii]] : attempt to select less than one element
```

Not sure what the is right `data.table` way of getting this: 

```
> iris %>% group_by(Species) %>% summarize
Source: local data frame [3 x 1]

     Species
1     setosa
2 versicolor
3  virginica
```

**Comment by krlmlr** [comment]
I'm seeing this with current `data.table` (Rdatatable/data.table@fb1b542d7a7fbf0072c0ad45c0d976f3c97aca0d):

```
> iris_dt[, list(), by = "Species"]
Empty data.table (0 rows) of 1 col: Species
```

Which doesn't exactly help here either. Of course you can do 

```
> iris_dt[, NA, by = "Species"]
      Species V1
1:     setosa NA
2: versicolor NA
3:  virginica NA
```

and remove the dummy column, but I think

```
> unique(iris_dt, by = "Species")
   Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
1:          5.1         3.5          1.4         0.2     setosa
2:          7.0         3.2          4.7         1.4 versicolor
3:          6.3         3.3          6.0         2.5  virginica
```

gives better results. I have yet to figure out how to create a data.table with only the `Species` column. 

CC @mdowle and @arunsrinivasan: Can suggest a better solution?

(I think it's fair to assume that corner cases like this will cause trouble with most backends. It would be great if dplyr "just worked" here, but this seems to require a bit of extra effort.)

**Comment by arunsrinivasan** [comment]
```r
require(data.table) # v1.9.5
uniqueN(iris_dt, by = "Species")
# [1] 3
```
Also raised https://github.com/Rdatatable/data.table/issues/1269. Thanks.

**Comment by krlmlr** [comment]
@arunsrinivasan: Thanks for your fast reply. Actually, I added the `nrow` call only to shorten the output; from dplyr's perspective, we need to handle this part:

```
iris %>% tbl_dt %>% group_by(Species) %>% summarize
```

The data.table issue you created covers it, but I wonder if

```
> iris_dt <- as.data.table(iris)
> unique(iris_dt[, "Species", with = FALSE], by = "Species")
      Species
1:     setosa
2: versicolor
3:  virginica
```

would be a good way to handle the issue from dplyr's end. (Here, using the "by" for both the j and the by arguments is not a problem.)

@mattdowle: Did you receive the notification from my earlier comment? I mistyped your user name when posting, and corrected it later -- just curious how GitHub handles this.
